### PR TITLE
Universal/PPv2 compatibility layer

### PR DESCRIPTION
**Pull request by Chman**
### Purpose of this PR

Added an option to temporarily support PPv2 in URP to ease the porting of LW projects to URP. It's a project-wide setting, meaning you can't use builtin & ppv2 effects at the same time.

When the PPv2 package is installed, you'll find a new option in the URP Asset:

image

That this option is hidden and reverted to "Integrated" when the package is uninstalled.

Note: #5462 has been merged into this PR in preparation for the final merge so it's probably best to wait for it to be merged into `release/2019.3` before reviewing or testing it.

### CODE REVIEW

File: com.unity.render-pipelines.universal/Editor/Overrides/TonemappingEditor.cs [hunk 1]
```diff
@@ -16,15 +16,17 @@ public override void OnEnable()
...
+
             PropertyField(m_Mode);
 
             // Display a warning if the user is trying to use a tonemap while rendering in LDR
-            var asset = UniversalRenderPipeline.asset;
-            if (asset != null && !asset.supportsHDR)
-            {
+            if (UniversalRenderPipeline.asset?.supportsHDR == false)
                 EditorGUILayout.HelpBox("Tonemapping should only be used when working in HDR.", MessageType.Warning);
-                return;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/Overrides/TonemappingEditor.cs]: is this removal intentional?

File: com.unity.render-pipelines.universal/Editor/Unity.RenderPipelines.Universal.Editor.asmdef [hunk 2]
```diff
@@ -17,5 +17,11 @@
     "precompiledReferences": [],
     "autoReferenced": true,
     "defineConstraints": [],
-    "versionDefines": []
-}
\ No newline at end of file
+    "versionDefines": [
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/Unity.RenderPipelines.Universal.Editor.asmdef]: We should  not add hard dependency, instead we force ppv3 if ppv2 is not present.

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 3]
```diff
@@ -55,11 +55,15 @@ internal class Styles
             public static GUIContent supportsSoftShadows = EditorGUIUtility.TrTextContent("Soft Shadows", "If enabled pipeline will perform shadow filtering. Otherwise all lights that cast shadows will fallback to perform a single shadow sample.");
 
             // Post-processing
+            public static GUIContent postProcessingFeatureSet = EditorGUIUtility.TrTextContent("Feature Set", "Sets the feature set to use for the render pipeline. Builtin should always be preferred over the other compatibility mode.");
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: What name are we using in docs for PPv3? I'm worried builtin will cause confusion. Maybe we should use `Integrated post-processing stack`?

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 4]
```diff
@@ -55,11 +55,15 @@ internal class Styles
...
             public static GUIContent colorGradingMode = EditorGUIUtility.TrTextContent("Grading Mode", "Defines how color grading will be applied. Operators will react differently depending on the mode.");
             public static GUIContent colorGradingLutSize = EditorGUIUtility.TrTextContent("LUT size", "Sets the size of the internal and external color grading lookup textures (LUTs).");
+            public static string postProcessingFeatureSetWarning = "Post-processing V2 support for the Universal Render Pipeline will be deprecated in a near future. We recommend only using this mode for compatibility purpose.";
+            public static string postProcessingFeatureSetError = "The Post-processing V2 package isn't installed, the feature set will fallback on Builtin.";
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: Integrated? PPv3?

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 5]
```diff
@@ -345,16 +351,32 @@ void DrawPostProcessingSettings()
...
-                else if (isHdrOn && m_ColorGradingMode.intValue == (int)ColorGradingMode.HighDynamicRange)
-                    EditorGUILayout.HelpBox(Styles.colorGradingModeSpecInfo, MessageType.Info);
+                EditorGUILayout.PropertyField(m_PostProcessingFeatureSet, Styles.postProcessingFeatureSet);
 
-                EditorGUILayout.DelayedIntField(m_ColorGradingLutSize, Styles.colorGradingLutSize);
-                m_ColorGradingLutSize.intValue = Mathf.Clamp(m_ColorGradingLutSize.intValue, UniversalRenderPipelineAsset.k_MinLutSize, UniversalRenderPipelineAsset.k_MaxLutSize);
-                if (isHdrOn && m_ColorGradingMode.intValue == (int)ColorGradingMode.HighDynamicRange && m_ColorGradingLutSize.intValue < 32)
-                    EditorGUILayout.HelpBox(Styles.colorGradingLutSizeWarning, MessageType.Warning);
+                bool ppv2Enabled = false;
+                if (m_PostProcessingFeatureSet.intValue == (int)PostProcessingFeatureSet.PostProcessingV2)
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: Maybe we should wrap the `POST_PROCESSING_STACK_2_0_0_OR_NEWER` define in this if as well. We should force ppv3/integrated when no ppv2 package is installed instead of having it as hard dependency. 

This is also a good way to know if we can "upgrade" a project straight away to v3.

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs [hunk 6]
```diff
@@ -12,6 +12,10 @@ internal class Renderer2D : ScriptableRenderer
         FinalBlitPass m_FinalBlitPass;
         PostProcessPass m_FinalPostProcessPass;
 
+#if POST_PROCESSING_STACK_2_0_0_OR_NEWER
+        PostProcessPassCompat m_PostProcessPassCompat;
+#endif
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs]: Was PPv2 supported in 2D renderer before? I wonder if we should add compatibility for it.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 7]
```diff
@@ -98,6 +98,12 @@ public enum ColorGradingMode
         HighDynamicRange
     }
 
+    public enum PostProcessingFeatureSet
+    {
+        Builtin,
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs]: Integrated/v3?

File: com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs [hunk 8]
```diff
@@ -221,60 +243,106 @@ public override void Setup(ScriptableRenderContext context, ref RenderingData re
             bool requiresFinalPostProcessPass = postProcessEnabled &&
                                      renderingData.cameraData.antialiasing == AntialiasingMode.FastApproximateAntialiasing;
 
-            // if we have additional filters
-            // we need to stay in a RT
-            if (afterRenderExists)
+#if POST_PROCESSING_STACK_2_0_0_OR_NEWER
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs]: Add a #region here so we can fold out this easily.

File: com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs [hunk 9]
```diff
@@ -221,60 +243,106 @@ public override void Setup(ScriptableRenderContext context, ref RenderingData re
...
+                // we need to stay in a RT
+                if (afterRenderExists)
                 {
-                    if (requiresFinalPostProcessPass)
+                    bool willRenderFinalPass = (m_ActiveCameraColorAttachment != RenderTargetHandle.CameraTarget);
+                    // perform post with src / dest the same
+                    if (postProcessEnabled)
                     {
-                        m_PostProcessPass.Setup(cameraTargetDescriptor, m_ActiveCameraColorAttachment, m_AfterPostProcessColor, m_ActiveCameraDepthAttachment, m_ColorGradingLut, true, false);
+                        m_PostProcessPass.Setup(cameraTargetDescriptor, m_ActiveCameraColorAttachment, m_AfterPostProcessColor, m_ActiveCameraDepthAttachment, m_ColorGradingLut, requiresFinalPostProcessPass, !willRenderFinalPass);
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs]: Seems like some additional changes in PostProcessing not just related to ppv2? Is it fixing additional issue? If so we should make a separate PR to master if we need those as this one will only be available in 2019.3

File: com.unity.render-pipelines.universal/Editor/Overrides/TonemappingEditor.cs (hunk 1 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Editor/Overrides/TonemappingEditor.cs]: Yep, it's at the end of the function so it's redundant :)

File: com.unity.render-pipelines.universal/Editor/Unity.RenderPipelines.Universal.Editor.asmdef (hunk 2 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Editor/Unity.RenderPipelines.Universal.Editor.asmdef]: I'm not sure I understand, this is not a dependency, it's just a define set if the PPv2 package is installed. The same is done in the Runtime asmdef, I just copied it over to the Editor asmdef because I needed it there as well.

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs (hunk 3 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: Yeah I wasn't sure to be honest. This is where I'd like to get some input from doc people :)

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs (hunk 6 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs]: It was, I added it back for the sake of consistency.

File: com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs (hunk 8 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs]: I can add a region, but as my IDE supports folding out #if/#else I didn't think it would be necessary.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 7 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs]: Integrated works for me. No v3 though, as it would make it confusing (it's not a package anymore, and "v3" in HDRP isn't the same either).

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs (hunk 5 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: I'm still not sure what you mean about having a hard dependency, there isn't any. Do you mean that we should just hide everything in the UI if PPv2 isn't installed?

File: com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs (hunk 9 above)
Comment (Chman) [reply com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs]: Hmm right, it's part of a fix that reached master but isn't in 2019.3. I'll remove it.

Comment (Chman) [reply com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs]: Turns out it's not a change, it's the diff tool failing at figuring out which line was actually modified. You can check the source files manually, the code hasn't changed.

**Comment by Chman** [comment]
For the doc I will have to add a section about PPv2 compatibility in the final PR (this one is just a draft). In the meantime, can we check that the 4 messages/tooltips added in `UniversalRenderPipelineAssetEditor.cs` are good?

File: com.unity.render-pipelines.universal/Editor/Unity.RenderPipelines.Universal.Editor.asmdef (hunk 2 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Editor/Unity.RenderPipelines.Universal.Editor.asmdef]: Ah ok. I misunderstood it.

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 10]
```diff
@@ -55,11 +55,15 @@ internal class Styles
             public static GUIContent supportsSoftShadows = EditorGUIUtility.TrTextContent("Soft Shadows", "If enabled pipeline will perform shadow filtering. Otherwise all lights that cast shadows will fallback to perform a single shadow sample.");
 
             // Post-processing
+            public static GUIContent postProcessingFeatureSet = EditorGUIUtility.TrTextContent("Feature Set", "Sets the feature set to use for the render pipeline. Integrated should always be preferred over the other compatibility mode.");
             public static GUIContent colorGradingMode = EditorGUIUtility.TrTextContent("Grading Mode", "Defines how color grading will be applied. Operators will react differently depending on the mode.");
             public static GUIContent colorGradingLutSize = EditorGUIUtility.TrTextContent("LUT size", "Sets the size of the internal and external color grading lookup textures (LUTs).");
+            public static string postProcessingFeatureSetWarning = "Post-processing V2 support for the Universal Render Pipeline will be deprecated in a near future. We recommend only using this mode for compatibility purpose.";
```
Comment (JordanL8) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: Unity plans to deprecate Post-processing V2 support for the Universal Render Pipeline in the near future. You should only use this mode for compatibility purposes.

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 11]
```diff
@@ -55,11 +55,15 @@ internal class Styles
             public static GUIContent supportsSoftShadows = EditorGUIUtility.TrTextContent("Soft Shadows", "If enabled pipeline will perform shadow filtering. Otherwise all lights that cast shadows will fallback to perform a single shadow sample.");
 
             // Post-processing
+            public static GUIContent postProcessingFeatureSet = EditorGUIUtility.TrTextContent("Feature Set", "Sets the feature set to use for the render pipeline. Integrated should always be preferred over the other compatibility mode.");
```
Comment (JordanL8) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: Sets the feature set to use for the render pipeline. To future proof your application, use Integrated instead of the comparability mode. Only use compatibility mode if your project still uses the Post-processing V2 package, but be aware that Unity plans to deprecate Post-processing V2 support for the Universal Render Pipeline in the near future.

This is a bit big for a tooltip, but I think its quite crucial information.

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 12]
```diff
@@ -55,11 +55,15 @@ internal class Styles
...
             public static GUIContent colorGradingMode = EditorGUIUtility.TrTextContent("Grading Mode", "Defines how color grading will be applied. Operators will react differently depending on the mode.");
             public static GUIContent colorGradingLutSize = EditorGUIUtility.TrTextContent("LUT size", "Sets the size of the internal and external color grading lookup textures (LUTs).");
+            public static string postProcessingFeatureSetWarning = "Post-processing V2 support for the Universal Render Pipeline will be deprecated in a near future. We recommend only using this mode for compatibility purpose.";
+            public static string postProcessingFeatureSetError = "The Post-processing V2 package isn't installed, the feature set will fallback on Integrated.";
```
Comment (JordanL8) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: Your Project does not include the Post-processing V2 package. The Universal Render Pipeline will use its Integrated post-processing solution.

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 13]
```diff
@@ -55,11 +55,15 @@ internal class Styles
...
             public static string colorGradingModeWarning = "HDR rendering is required to use the high dynamic range color grading mode. The low dynamic range will be used instead.";
             public static string colorGradingModeSpecInfo = "The high dynamic range color grading mode works best on platforms that support floating point textures.";
             public static string colorGradingLutSizeWarning = "The minimal recommended LUT size for the high dynamic range color grading mode is 32. Using lower values will potentially result in color banding and posterization effects.";
+            public static string postProcessingGlobalWarning = "The Post-processing Feature Set in the render pipeline asset is set to Post-processing V2. This volume component will not have any effect.";
```
Comment (JordanL8) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs]: The Post-processing Feature Set in the URP Asset is set to Post-processing V2. This Volume component will not have any effect.

**Comment by JordanL8** [comment]
Tooltips look good to me. Also, I agree that we should use 'integrated' for all the docs and tooltips in the future.

**Comment by Chman** [comment]
Fixed, thanks for the input!

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 14]
```diff
@@ -98,6 +98,12 @@ public enum ColorGradingMode
         HighDynamicRange
     }
 
+    public enum PostProcessingFeatureSet
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs]: Either make this internal or add API docs to any new public API.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 15]
```diff
@@ -583,6 +592,26 @@ public bool useSRPBatcher
             set { m_UseSRPBatcher = value; }
         }
 
+        public PostProcessingFeatureSet postProcessingFeatureSet
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs]: Either make this internal or add API docs to any new public API.

File: com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs [hunk 16]
```diff
@@ -244,60 +255,111 @@ public override void Setup(ScriptableRenderContext context, ref RenderingData re
...
+                // we need to stay in a RT
+                if (afterRenderExists)
                 {
-                    if (renderingData.cameraData.captureActions != null)
+                    // perform post with src / dest the same
+                    if (applyPostProcessing)
                     {
-                        m_CapturePass.Setup(m_ActiveCameraColorAttachment);
-                        EnqueuePass(m_CapturePass);
+                        m_PostProcessPassCompat.Setup(cameraTargetDescriptor, m_ActiveCameraColorAttachment, m_ActiveCameraColorAttachment);
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs]: If there's exactly one effect in the stack this will fail iirc.

Comment (Chman) [reply com.unity.render-pipelines.universal/Runtime/ForwardRenderer.cs]: This will work fine here, it's handled by PPv2. But saying this actually made me realize this PR was missing the support for opaque-only PPv2 effect, I added it back (and yeah, if there's only one opaque-only effect I believe it will fail as it did on LWRP)...
